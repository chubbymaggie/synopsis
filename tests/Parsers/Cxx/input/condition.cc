// Tests the ability to use a declaration in the condition of an if or switch
struct X
{
  operator bool() const {return true;}
  struct Y
  {
    operator bool() const {return true;}
  };
};
void foo()
{
  int _i;
  X _x;
  X::Y _y;

  if (int i = 3) {}
  if (const int i = 3) {}
  if (const int* i = &_i) {}
  if (int* const i = &_i) {}
  
  if (X foo = _x) {}
  if (const X foo = _x) {}
  if (const X* foo = &_x) {}
  if (X* const foo = &_x) {}

  if (X::Y foo = _y) {}
  if (const X::Y foo = _y) {}
  if (const X::Y* foo = &_y) {}
  if (X::Y* const foo = &_y) {}

  switch (int i = 3) {}
  switch (const int i = 3) {}
  
}
