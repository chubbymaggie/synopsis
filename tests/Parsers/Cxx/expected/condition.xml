<?xml version='1.0' encoding='ISO-8859-1'?>
<ast>
 <declarations>
  <instance accessibility="0" class="Synopsis.AST.Class" file="Parsers/Cxx/input/condition.cc" language="C++" line="2" name="X" type="struct">
   <comments>
    <instance class="Synopsis.AST.Comment" line="0" suspect="0">
     <file>
      Parsers/Cxx/input/condition.cc
     </file>
     <text>
      // Tests the ability to use a declaration in the condition of an if or switch

     </text>
    </instance>
   </comments>
   <declarations>
    <instance accessibility="1" class="Synopsis.AST.Operation" file="Parsers/Cxx/input/condition.cc" language="C++" line="4" name="X.(bool)()const" realname="(bool)" type="member function">
     <returnType>
      <instance class="Synopsis.Type.Base" language="C++" name="bool"/>
     </returnType>
    </instance>
    <instance accessibility="1" class="Synopsis.AST.Class" file="Parsers/Cxx/input/condition.cc" language="C++" line="5" name="X.Y" type="struct">
     <declarations>
      <instance accessibility="1" class="Synopsis.AST.Operation" file="Parsers/Cxx/input/condition.cc" language="C++" line="7" name="X.Y.(bool)()const" realname="(bool)" type="member function">
       <returnType/>
      </instance>
      <instance accessibility="1" class="Synopsis.AST.Builtin" file="Parsers/Cxx/input/condition.cc" language="C++" line="7" name="EOS" type="EOS"/>
     </declarations>
    </instance>
    <instance accessibility="1" class="Synopsis.AST.Builtin" file="Parsers/Cxx/input/condition.cc" language="C++" line="7" name="EOS" type="EOS"/>
   </declarations>
  </instance>
  <instance accessibility="0" class="Synopsis.AST.Function" file="Parsers/Cxx/input/condition.cc" language="C++" line="10" name="foo()" realname="foo" type="function">
   <returnType>
    <instance class="Synopsis.Type.Base" language="C++" name="void"/>
   </returnType>
  </instance>
 </declarations>
 <types>
  <instance class="Synopsis.Type.Declared" language="C++" name="EOS">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="X">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="X.(bool)()const">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="X.Y">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="X.Y.(bool)()const">
   <declaration/>
  </instance>
  <instance class="Synopsis.Type.Declared" language="C++" name="foo()">
   <declaration/>
  </instance>
 </types>
 <files>
  <file>
   Parsers/Cxx/input/condition.cc
  </file>
 </files>
</ast>
